source/lac/scalapack.cc

```diff
@@ -673,19 +675,92 @@ void ScaLAPACKMatrix<NumberType>::compute_cholesky_factorization()
+
+  Threads::Mutex::ScopedLock lock (mutex);
+
+  if (grid->mpi_process_is_active)
+    {
+      int info = 0;
+      NumberType *A_loc = &this->values[0];
+
+      const int iarow = indxg2p_(&submatrix_row, &row_block_size, &(grid->this_process_row), &first_process_row, &(grid->n_process_rows));
+      const int mp = numroc_(&n_rows, &row_block_size, &(grid->this_process_row), &iarow, &(grid->n_process_rows));
+      ipiv.resize(mp+row_block_size);
+
+      pgetrf(&n_rows,&n_columns,A_loc,&submatrix_row,&submatrix_column,descriptor,ipiv.data(),&info);
+      AssertThrow (info==0, LAPACKSupport::ExcErrorCode("pgetrf", info));
+    }
+  state = LAPACKSupport::State::lu;
+}
+
+
+
 template <typename NumberType>
 void ScaLAPACKMatrix<NumberType>::invert()
 {
-  if (state == LAPACKSupport::matrix)
-    compute_cholesky_factorization();
+  // Check whether matrix is symmetric and save flag.
+  bool is_symmetric = (property == LAPACKSupport::symmetric) ? true : false;
+
+  // If a Cholesky factorization has been applied previously,
+  // the original matrix was symmetric
+  if (state == LAPACKSupport::State::cholesky)
+    is_symmetric = true;
+
+  // Matrix is neither in Cholesky nor LU state.
+  // Compute the required factorizations based on the property of the matrix.
+  if ( !(state==LAPACKSupport::State::lu || state==LAPACKSupport::State::cholesky) )
+    {
+      Assert (state == LAPACKSupport::matrix,
+              ExcMessage("Matrix has to be in Matrix state before calling this function."));
+
+      if (is_symmetric)
+        compute_cholesky_factorization();
+      else
+        compute_lu_factorization();
+    }
+
+  Threads::Mutex::ScopedLock lock (mutex);
 
   if (grid->mpi_process_is_active)
     {
       int info = 0;
       NumberType *A_loc = &this->values[0];
-      ppotri (&uplo,&n_columns, A_loc, &submatrix_row, &submatrix_column, descriptor,&info);
-      AssertThrow (info==0, LAPACKSupport::ExcErrorCode("ppotri", info));
+
+      if (is_symmetric)
+        {
+          ppotri(&uplo, &n_columns, A_loc, &submatrix_row, &submatrix_column, descriptor, &info);
+          AssertThrow (info==0, LAPACKSupport::ExcErrorCode("ppotri", info));
+        }
+      else
+        {
+          int lwork=-1, liwork=-1;
+          work.resize(1);
+          iwork.resize(1);
+
+          pgetri(&n_columns, A_loc, &submatrix_row, &submatrix_column, descriptor,
+                 ipiv.data(), work.data(), &lwork, iwork.data(), &liwork, &info);
+
+          AssertThrow (info==0, LAPACKSupport::ExcErrorCode("pgetri", info));
+          lwork=work[0];
+          liwork=iwork[0];
+          work.resize(lwork);
+          iwork.resize(liwork);
+
+          pgetri(&n_columns, A_loc, &submatrix_row, &submatrix_column, descriptor,
+                 ipiv.data(), work.data(), &lwork, iwork.data(), &liwork, &info);
+
+          AssertThrow (info==0, LAPACKSupport::ExcErrorCode("pgetri", info));
+        }
     }
+  if (is_symmetric)
```

Comment:
that's a duplicate of what we already do in `cholesky()`. I would remove both, and maybe set `propety` in `lu` function.